catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/metacard/workspace/transformer/impl/QueryTransformer.java

```diff
@@ -0,0 +1,129 @@
+import java.util.Objects;
+import java.util.stream.Collectors;
+import org.codice.ddf.catalog.ui.metacard.workspace.QueryMetacardImpl;
+import org.codice.ddf.catalog.ui.query.transform.QueryValueTransformer;
+import org.codice.ddf.catalog.ui.util.EndpointUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Utility class that performs common functions for handling the input and output of query endpoints
+ */
+public class QueryTransformer {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(QueryTransformer.class);
+
+  private EndpointUtil endpointUtil;
+
+  private List<QueryValueTransformer> queryValueTransformers;
+
+  private static final Map<String, String> FIELD_MAPPINGS = ImmutableMap.of("src", "sources");
+
+  public QueryTransformer(
+      EndpointUtil endpointUtil, List<QueryValueTransformer> queryValueTransformers) {
+    this.endpointUtil = endpointUtil;
+    this.queryValueTransformers = queryValueTransformers;
+
+    Collections.sort(queryValueTransformers);
+  }
+
+  public Map<String, Object> transformMetacardIntoMap(Metacard metacard) {
+    Map<String, Object> map = endpointUtil.getMetacardMap(metacard);
+
+    return map.entrySet()
+        .stream()
+        .filter(entry -> Objects.nonNull(entry.getValue()))
+        .filter(entry -> isBasicType(entry.getValue()))
+        .map(
+            entry -> {
+              String key = entry.getKey();
+              Object value = entry.getValue();
+
+              QueryValueTransformer transformer = getQueryValueTransformer(key, value);
+
+              if (transformer != null) {
+                return new AbstractMap.SimpleEntry<>(key, transformer.getJsonValue(key, value));
+              } else {
+                LOGGER.trace(
+                    "Appropriate query value transformer could not be found for [{}] with value [{}]",
+                    key,
+                    value);
+              }
+
+              return null;
+            })
+        .filter(Objects::nonNull)
+        .collect(Collectors.toMap(Entry::getKey, Entry::getValue));
+  }
+
+  public Metacard transformJsonIntoMetacard(Map<String, Object> json) {
+    QueryMetacardImpl query = new QueryMetacardImpl();
+
+    json.forEach(
+        (key, value) -> {
+          QueryValueTransformer transformer = getQueryValueTransformer(key, value);
+
+          if (transformer != null) {
+            String attrKey = FIELD_MAPPINGS.getOrDefault(key, key);
+
+            Attribute attr = transformer.createAttribute(attrKey, value);
+            query.setAttribute(attr);
+          } else {
+            LOGGER.trace(
+                "Appropriate query value transformer could not be found for [{}] with value [{}]",
+                key,
+                value);
+          }
+        });
+
+    return query;
+  }
+
+  private boolean isBasicType(Object value) {
+    return !(value instanceof List) || notEmptyList(value);
+  }
+
+  private boolean notEmptyList(Object value) {
```

Comment:
✏️  we could use the apache commons library since they have a very similar method. Also if we're only calling it from the method above, we might as well cast it to a List since we know its a list by then and we wont have to check a second time down in this method.

Reply:
This is a map representation of a metacard so not every entry's value is a list. This is what makes this operation about complex. We want to filer out `null` and empty values while leaving in everything else but we have to perform a check that the object is what not a list before we can check if it's empty. 

Any other solution is going to look very similar to this